// ox/_cjs/tempo/TempoAddress.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidChecksumError = exports.InvalidLengthError = exports.InvalidPrefixError = void 0;
exports.format = format;
exports.parse = parse;
exports.validate = validate;
const Address = require("../core/Address.js");
const Base32 = require("../core/Base32.js");
const Bytes = require("../core/Bytes.js");
const CompactSize = require("../core/CompactSize.js");
const Errors = require("../core/Errors.js");
const Hash = require("../core/Hash.js");
const Hex = require("../core/Hex.js");
function format(address, options = {}) {
    const { zoneId } = options;
    const prefix = zoneId != null ? 'tempoz1' : 'tempo1';
    const zone = zoneId != null ? CompactSize.toBytes(zoneId) : new Uint8Array();
    const address_bytes = Bytes.fromHex(address);
    const input = Bytes.concat(Bytes.fromString(prefix), zone, address_bytes);
    const checksum = Hash.sha256(Hash.sha256(input, { as: 'Bytes' }), {
        as: 'Bytes',
    }).slice(0, 4);
    const payload = Bytes.concat(zone, address_bytes, checksum);
    return `${prefix}${Base32.fromBytes(payload)}`;
}
function parse(tempoAddress) {
    const lower = tempoAddress.toLowerCase();
    let prefix;
    let hasZone;
    if (lower.startsWith('tempoz1')) {
        prefix = 'tempoz1';
        hasZone = true;
    }
    else if (lower.startsWith('tempo1')) {
        prefix = 'tempo1';
        hasZone = false;
    }
    else {
        throw new InvalidPrefixError({ address: tempoAddress });
    }
    const payload = Base32.toBytes(lower.slice(prefix.length));
    let zoneId;
    let remaining;
    if (hasZone) {
        const { value, size } = CompactSize.fromBytes(payload);
        zoneId = value;
        remaining = payload.slice(size);
    }
    else {
        zoneId = undefined;
        remaining = payload;
    }
    if (remaining.length !== 24)
        throw new InvalidLengthError({
            address: tempoAddress,
            expected: 24,
            actual: remaining.length,
        });
    const rawAddress = remaining.slice(0, 20);
    const checksum = remaining.slice(20, 24);
    const zoneBytes = zoneId != null ? CompactSize.toBytes(zoneId) : new Uint8Array();
    const checksumInput = Bytes.concat(Bytes.fromString(prefix), zoneBytes, rawAddress);
    const expected = Hash.sha256(Hash.sha256(checksumInput, { as: 'Bytes' }), {
        as: 'Bytes',
    }).slice(0, 4);
    if (!Bytes.isEqual(checksum, expected))
        throw new InvalidChecksumError({ address: tempoAddress });
    const address = Address.checksum(Hex.fromBytes(rawAddress));
    return { address, zoneId };
}
function validate(tempoAddress) {
    try {
        parse(tempoAddress);
        return true;
    }
    catch {
        return false;
    }
}
class InvalidPrefixError extends Errors.BaseError {
    constructor({ address }) {
        super(`Tempo address "${address}" has an invalid prefix. Expected "tempo1" or "tempoz1".`);
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 'TempoAddress.InvalidPrefixError'
        });
    }
}
exports.InvalidPrefixError = InvalidPrefixError;
class InvalidLengthError extends Errors.BaseError {
    constructor({ address, expected, actual, }) {
        super(`Tempo address "${address}" has an invalid payload length. Expected ${expected} bytes, got ${actual}.`);
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 'TempoAddress.InvalidLengthError'
        });
    }
}
exports.InvalidLengthError = InvalidLengthError;
class InvalidChecksumError extends Errors.BaseError {
    constructor({ address }) {
        super(`Tempo address "${address}" has an invalid checksum.`);
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 'TempoAddress.InvalidChecksumError'
        });
    }
}
exports.InvalidChecksumError = InvalidChecksumError;
